ledger/participant-state/src/main/scala/com/daml/ledger/participant/state/v2/WriteService.scala

```diff
@@ -0,0 +1,155 @@
+    * @param transaction                 the submitted transaction. This transaction can contain local
+    *                                    contract-ids that need suffixing. The participant state may have to
+    *                                    suffix those contract-ids in order to guaranteed their global
+    *                                    uniqueness. See the Contract Id specification for more detail
+    *                                    daml-lf/spec/contract-id.rst.
+    * @param estimatedInterpretationCost Estimated cost of interpretation that may be used for
+    *                                    handling submitted transactions differently.
+    * @param telemetryContext            Implicit context for tracing.
+    */
+  def submitTransaction(
+      submitterInfo: SubmitterInfo,
+      transactionMeta: TransactionMeta,
+      transaction: SubmittedTransaction,
+      estimatedInterpretationCost: Long,
+  )(implicit telemetryContext: TelemetryContext): CompletionStage[SubmissionResult]
+
+  /** Record the rejection of a command submission on the ledger.
+    *
+    * TODO(v2) Why do we need the requirement below on no IO? Canton doesn't respect it at the moment.
+    *  For command deduplication, we will need to do some IO inside the write service,
+    *  and want to report the outcome synchronously whenever possible.
+    *
+    * This method must be thread-safe, not throw, and not block on IO. It is
+    * though allowed to perform significant computation. The expectation is
+    * that callers call this method on a thread dedicated to getting this transaction
+    * ready for acceptance to the ledger, which typically requires some
+    * preparation steps (decomposition, serialization) by the implementation
+    * of the [[WriteService]].
+    *
+    * The result is communicated asynchronously via a [[ReadService]] implementation backed by the same participant
+    * state as this [[WriteService]]. Successful recording is communicated using a [[Update.CommandRejected]]
+    * with [[SubmitterInfo]] and not [[Update.CommandRejected.cancelled]].
+    * If the recording as a rejection fails (e.g., due to deduplication or violations of the submission rank),
+    * the failure should be communicated using a [[Update.CommandRejected]] with [[SubmitterInfo]]
+    * and [[Update.CommandRejected.cancelled]].
+    *
+    * Recorded rejections fall under the deduplication and submission rank guarantees
+    * described in [[ReadService.stateUpdates]].
+    *
+    * @param submitterInfo the information provided by the submitter for correlating this submission
+    *                      with its rejection or cancellation on the associated [[ReadService]].
+    * @param reason The rejection reason to be included in the [[Update.CommandRejected]]
+    * @param telemetryContext Implicit context for tracing.
+    */
+  def rejectSubmission(
+      submitterInfo: SubmitterInfo,
+      reason: com.google.rpc.Status
```

Comment:
This seems like the right approach. Do you expect the error details to be serialized in the IndexDB?

Reply:
Yes. See my proposal in the Ledger API design PR #9567.